src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs

```diff
@@ -618,6 +619,136 @@ public sealed class SocketsHttpHandler_ConnectionUpgrade_Test : HttpClientTestBa
         }
     }
 
+    public sealed class SocketsHttpHandler_Connect_Test : HttpClientTestBase
+    {
+        protected override bool UseSocketsHttpHandler => true;
+
+        [Fact]
+        public async Task ConnectMethod_Success()
+        {
+            await LoopbackServer.CreateServerAsync(async (server, url) =>
+            {
+                using (HttpClient client = CreateHttpClient())
+                {
+                    HttpRequestMessage request = new HttpRequestMessage(new HttpMethod("CONNECT"), url);
+                    request.Headers.Add("Host", String.Format("{0}:{1}", request.RequestUri.IdnHost, request.RequestUri.Port));
+
+                    // We need to use ResponseHeadersRead here, otherwise we will hang trying to buffer the response body.
+                    Task<HttpResponseMessage> responseTask = client.SendAsync(request,  HttpCompletionOption.ResponseHeadersRead);
+
+                    await server.AcceptConnectionAsync(async connection =>
+                    {
+                        // Verify that Host header exist and has same value and URI authority.
+                        List<string> lines = await connection.ReadRequestHeaderAsync().ConfigureAwait(false);
+                        string authority = lines[0].Split()[1];
+                        string hostHeaderValue = null;
```

Comment:
If you just hardwire a hostname above, you can simplify this by just verifying that there's a header that looks exactly like "Host: foo.com:345".

In general we can keep tests simple this way; we don't need a generic server with a generic header parser, we just need to verify exactly what we expected to send.